pages/blog/miniseries-part_1.md

```diff
@@ -0,0 +1,234 @@
+
+<figure>
+  <img src="/img/posts/jonaslagoni-miniseries-part1/blog-miniseries-processor-api.webp" title="Processor setup"/>
+  <figcaption className="text-center text-gray-400 text-sm">Displays the processor API as it is described with AsyncAPI with version 2.0.0. The round dot between "some broker" and the processor represent how others may grab/provide events that the processor subscribes to.</figcaption>
+</figure>
+This is again because I need to define how others may interact with our **processor**. This means that instead of using the `subscribe` operation I use `publish` to tell others that they can publish to this channel since the backend **processor** is subscribing to it. The full AsyncAPI document for the **processor** can be found [here](https://github.com/jonaslagoni/asyncapi-miniseries/blob/master/AsyncAPI/Processor.yaml). 
+
+```yaml
+...
+  game/server/{serverId}/events/player/{playerId}/item/{itemId}/pickup:
+    ...
+    publish: 
+      ...
+...
+```
+
+## Introducing reusability
+At the moment each of the AsyncAPI documents contains its own definition of the channels. But what if I were to add a new validation rule such as a new property to the **playerItemPickupPayload** schema? In this case, I would have to change this for both applications which is way too much work :smile:
+
+We can therefore introduce **$ref** to separate the parameters and messages into smaller sections for reusability. I will be placing all separate components into a ["components" directory](https://github.com/jonaslagoni/asyncapi-miniseries/tree/master/AsyncAPI) in the same directory the AsyncAPI documents reside.
+
+Just a quick note, at the moment it is not possible to reuse channels and operations directly between the two applications, therefore we can only apply this to the parameters and message individually while keeping some duplicate information \[[2](#channel-reusability)\]. 
+
+First I separate the different parameters, for simplicity I add all of them into the same file `./components/Parameters.yaml`.
+```yaml
+serverId:
+  description: The id of the server
+  schema: 
+    type: string
+playerId:
+  description: The id of the player who performed the action
+  schema: 
+    type: string
+itemId:
+  description: The id of the item
+  schema: 
+    type: string
+```
+
+And then change all the channel parameters to reference the external parameter definition.
+```yaml
+...
+  game/server/{serverId}/events/player/{playerId}/item/{itemId}/pickup:
+    description: Channel used when a player picks up an item in-game
+    parameters:
+      serverId: 
+        $ref: "./components/Parameters.yaml#/serverId"
+      playerId: 
+        $ref: "./components/Parameters.yaml#/playerId"
+      itemId: 
+        $ref: "./components/Parameters.yaml#/itemId"
+    ...
+...
+```
+
+For the messages I add a new file per message instead of keeping everything in the same file as parameters, I use this approach since I find it easier to maintain and extend.
```

Comment:
```suggestion
For the messages, I add a new file per message instead of keeping everything in the same file as parameters. I use this approach since I find it easier to maintain and extend.
```